### PR TITLE
Reduce memory use in h evaluation

### PR DESCRIPTION
**Pull request by Brechtpd**
Removes the memory allocation needed previously needed for storing intermediate lookup values. This was necessary because all expressions (custom constraints and lookups) were added to the same graph so shared intermediate results could be shared. This was faster for the EVM circuit especially because input expressions for lookups were complicated. However this is not the case anymore and generally this will be pretty unlikely to help much in circuits because the EVM circuit was pretty special in that aspect. To keep things simple I chose to remove the previous behavior instead of something more complicated like being able to switch between both approaches.

### CODE REVIEW

File: halo2_proofs/src/plonk/evaluation.rs [hunk 1]
```diff
@@ -701,24 +493,232 @@ impl<C: CurveAffine> Evaluator<C> {
...
-#[allow(unsafe_code)]
-unsafe impl<T> Send for ThreadBox<T> {}
-#[allow(unsafe_code)]
-unsafe impl<T> Sync for ThreadBox<T> {}
-
-/// Wraps a mutable slice so it can be passed into a thread without
-/// hard to fix borrow checks caused by difficult data access patterns.
-impl<T> ThreadBox<T> {
-    fn wrap(data: &mut [T]) -> Self {
-        Self(data.as_mut_ptr(), data.len())
```
Comment (CPerezz) [inline halo2_proofs/src/plonk/evaluation.rs:715]: Feels nice to get ridd of this! :)

**Review by CPerezz — APPROVED** [review]
LGTM! 
Let me request a review to @han0110 and @kilic to make sure I haven't missed anything!

File: halo2_proofs/src/plonk/evaluation.rs [hunk 2]
```diff
@@ -701,24 +493,232 @@ impl<C: CurveAffine> Evaluator<C> {
...
+                }
+            },
+            Expression::Sum(a, b) => {
+                // Undo subtraction stored as a + (-b) in expressions
+                match &**b {
+                    Expression::Negated(b_int) => {
+                        let result_a = self.add_expression(a);
+                        let result_b = self.add_expression(b_int);
+                        if result_a == ValueSource::Constant(0) {
+                            result_b
```
Comment (han0110) [inline halo2_proofs/src/plonk/evaluation.rs:615]: Need to update here

```suggestion
                            self.add_calculation(Calculation::Negate(result_b))
```

**Review by han0110 — COMMENTED** [review]
LGTM! Need a rebase with update about calculation.

**Comment by Brechtpd** [comment]
Rebased!

**Review by han0110 — APPROVED** [review]

**Pull request merged**